### PR TITLE
add cart merge for anonymous cart

### PR DESCRIPTION
**Pull request by kaka89**

### CODE REVIEW

File: src/main/java/io/reactivesw/cart/application/controller/CartController.java [hunk 1]
```diff
@@ -92,13 +95,14 @@ public CartView checkout(@PathVariable(Router.CART_ID) String cartId) {
    * @return the cart by customer id
    */
   @GetMapping(value = Router.CARTS_ROOT, params = "customerId")
-  public CartView getActiveCartByCustomerId(@RequestParam @NotNull String customerId) {
-    LOG.info("customerId : {}", customerId);
+  public CartView getActiveCartByCustomerId(@RequestParam @NotNull String customerId,
```
Comment (davis999) [inline src/main/java/io/reactivesw/cart/application/controller/CartController.java:98]: getActiveCartByCustomerId is not appropriate, getActiveCart will be better.

File: src/main/java/io/reactivesw/cart/application/service/CartApplication.java [hunk 2]
```diff
@@ -71,17 +77,58 @@ public CartView updateCart(String id, Integer version, List<UpdateAction> action
...
+  public CartView getCartById(String id, String anonymousId) {
+    LOG.debug("Enter. id: {}, anonymousId: {}.", id, anonymousId);
     Cart entity = cartService.getById(id);
+
+    if (anonymousId != null) {
+      Cart anonymousCart = cartService.getCartByAnonymousId(anonymousId);
+
+      if (anonymousCart != null
+          && anonymousCart.getLineItems() != null
+          && anonymousCart.getLineItems().size() > 0) {
```
Comment (davis999) [inline src/main/java/io/reactivesw/cart/application/service/CartApplication.java]: ! isEmpty may be better.

File: src/main/java/io/reactivesw/cart/application/controller/CartController.java [hunk 3]
```diff
@@ -92,13 +95,14 @@ public CartView checkout(@PathVariable(Router.CART_ID) String cartId) {
...
+                                            @RequestParam(required = false) String anonymousId) {
+    LOG.info("Enter. customerId : {}, anonymousId: {}.", customerId);
 
-    Cart entity = cartService.getActiveCartByCustomerId(customerId);
+    CartView cartView = cartApplication.getCartByCustomerId(customerId, anonymousId);
 
-    LOG.info("entity : {}", entity);
-    return cartApplication.getFullCart(entity);
+    LOG.info("Exit. cartView : {}", cartView);
```
Comment (davis999) [inline src/main/java/io/reactivesw/cart/application/controller/CartController.java]: need a dot.
cartView should be CartView.

File: src/main/java/io/reactivesw/cart/application/service/CartApplication.java [hunk 4]
```diff
@@ -71,17 +77,58 @@ public CartView updateCart(String id, Integer version, List<UpdateAction> action
...
+        cartMerger.mergeCart(entity, anonymousCart);
+      } else {
+        LOG.debug("Anonymous cart not exist. anonymousId: {}", anonymousId);
+      }
+    }
+    LOG.debug("Exit. cart: {}", entity);
```
Comment (davis999) [inline src/main/java/io/reactivesw/cart/application/service/CartApplication.java]: need a dot.

File: src/main/java/io/reactivesw/cart/application/service/CartApplication.java [hunk 5]
```diff
@@ -71,17 +77,58 @@ public CartView updateCart(String id, Integer version, List<UpdateAction> action
...
 
   /**
+   * Get cart by customer id and anonymousId, if anonymous cart exist, then merge it to customer
+   * cart.
+   *
+   * @param customerId  String
+   * @param anonymousId String
+   * @return CartView
+   */
+  public CartView getCartByCustomerId(String customerId, String anonymousId) {
```
Comment (davis999) [inline src/main/java/io/reactivesw/cart/application/service/CartApplication.java:111]: I think getCartByCustomerId is not appropriate.

File: src/main/java/io/reactivesw/cart/application/service/CartApplication.java [hunk 6]
```diff
@@ -71,17 +77,58 @@ public CartView updateCart(String id, Integer version, List<UpdateAction> action
...
+    Cart customerCart = cartService.getActiveCartByCustomerId(customerId);
+
+    if (anonymousId != null) {
+      Cart anonymousCart = cartService.getCartByAnonymousId(anonymousId);
+      if (anonymousCart != null
+          && anonymousCart.getLineItems() != null
+          && anonymousCart.getLineItems().size() > 0) {
+        cartMerger.mergeCart(customerCart, anonymousCart);
+      } else {
+        LOG.debug("Anonymous cart not exist. anonymousId: {}", anonymousId);
```
Comment (davis999) [inline src/main/java/io/reactivesw/cart/application/service/CartApplication.java]: need a dot.

File: src/main/java/io/reactivesw/cart/application/service/CartApplication.java [hunk 7]
```diff
@@ -71,17 +77,58 @@ public CartView updateCart(String id, Integer version, List<UpdateAction> action
...
+   */
+  public CartView getCartByCustomerId(String customerId, String anonymousId) {
+    LOG.debug("Enter. customerId: {}, anonymousId: {}.", customerId, anonymousId);
+    Cart customerCart = cartService.getActiveCartByCustomerId(customerId);
+
+    if (anonymousId != null) {
+      Cart anonymousCart = cartService.getCartByAnonymousId(anonymousId);
+      if (anonymousCart != null
+          && anonymousCart.getLineItems() != null
+          && anonymousCart.getLineItems().size() > 0) {
```
Comment (davis999) [inline src/main/java/io/reactivesw/cart/application/service/CartApplication.java]: ! isEmpty may be better.

File: src/main/java/io/reactivesw/cart/application/service/CartMerger.java [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+  @Autowired
+  private transient AddLineItemService addLineItemService;
+
+  /**
+   * Merge cart.
+   *
+   * @param customerId  String
+   * @param anonymousId String.
+   */
+  public void mergeCart(String customerId, String anonymousId) {
```
Comment (davis999) [inline src/main/java/io/reactivesw/cart/application/service/CartMerger.java:45]: Does this method is useless?
I can find the usage.

File: src/main/java/io/reactivesw/cart/domain/service/CartService.java [hunk 9]
```diff
@@ -78,6 +78,23 @@ public Cart createActiveCart(Cart cart) {
   }
 
   /**
+   * Save cart.
+   *
+   * @param cart Cart
+   * @return Saved Cart
+   */
+  public Cart saveCart(Cart cart) {
+    LOG.debug("Enter. cart: {}.", cart);
```
Comment (davis999) [inline src/main/java/io/reactivesw/cart/domain/service/CartService.java:87]: cart should be Cart.

File: src/main/java/io/reactivesw/cart/domain/service/CartService.java [hunk 10]
```diff
@@ -78,6 +78,23 @@ public Cart createActiveCart(Cart cart) {
...
+    Cart result = null;
+    if (cart != null) {
+      result = cartRepository.save(cart);
+    }
+
+    LOG.debug("Exit. cart: {}.", result);
```
Comment (davis999) [inline src/main/java/io/reactivesw/cart/domain/service/CartService.java:93]: cart should be Cart.

File: src/main/java/io/reactivesw/cart/application/controller/CartController.java [hunk 11]
```diff
@@ -56,22 +56,25 @@ public CartController(CartService cartService, CartApplication cartApplication)
...
+   * @param cartId      the id
+   * @param anonymousId String
    * @return the cart by id
    */
   @GetMapping(Router.CART_WITH_ID)
-  public CartView getCartById(@PathVariable(Router.CART_ID) String cartId) {
-    LOG.info("Enter, id: {}", cartId);
+  public CartView getCartById(@PathVariable(Router.CART_ID) String cartId,
+                              @RequestParam(required = false) String anonymousId) {
+    LOG.info("Enter, id: {}, anonymousId: {}.", cartId, anonymousId);
```
Comment (davis999) [inline src/main/java/io/reactivesw/cart/application/controller/CartController.java:66]: should be "Enter. Id: {} ......"

File: src/main/java/io/reactivesw/cart/application/controller/CartController.java [hunk 12]
```diff
@@ -56,22 +56,25 @@ public CartController(CartService cartService, CartApplication cartApplication)
...
 
-    Cart entity = cartService.getById(cartId);
+    CartView cartView = cartApplication.getCartById(cartId, anonymousId);
 
-    LOG.info("Exit entity: {}", entity);
-    return cartApplication.getFullCart(entity);
+    LOG.info("Exit. cartView: {}.", cartView);
```
Comment (davis999) [inline src/main/java/io/reactivesw/cart/application/controller/CartController.java:70]: cartView should be CartView

File: src/main/java/io/reactivesw/cart/application/controller/CartController.java [hunk 13]
```diff
@@ -92,13 +95,14 @@ public CartView checkout(@PathVariable(Router.CART_ID) String cartId) {
    * @return the cart by customer id
    */
   @GetMapping(value = Router.CARTS_ROOT, params = "customerId")
-  public CartView getActiveCartByCustomerId(@RequestParam @NotNull String customerId) {
-    LOG.info("customerId : {}", customerId);
+  public CartView getActiveCartByCustomerId(@RequestParam @NotNull String customerId,
+                                            @RequestParam(required = false) String anonymousId) {
+    LOG.info("Enter. customerId : {}, anonymousId: {}.", customerId);
```
Comment (davis999) [inline src/main/java/io/reactivesw/cart/application/controller/CartController.java]: customerId should be CustomerId, and remove the useless space.

File: src/main/java/io/reactivesw/cart/application/service/CartApplication.java [hunk 14]
```diff
@@ -71,17 +77,58 @@ public CartView updateCart(String id, Integer version, List<UpdateAction> action
   }
 
   /**
-   * get cart by cart id.
+   * get cart by cart id, and if there is an anonymous id, then try to merge the anonymous cart
```
Comment (davis999) [inline src/main/java/io/reactivesw/cart/application/service/CartApplication.java]: uppercase letter first.

File: src/main/java/io/reactivesw/cart/application/service/CartApplication.java [hunk 15]
```diff
@@ -71,17 +77,58 @@ public CartView updateCart(String id, Integer version, List<UpdateAction> action
...
+
+    if (anonymousId != null) {
+      Cart anonymousCart = cartService.getCartByAnonymousId(anonymousId);
+
+      if (anonymousCart != null
+          && anonymousCart.getLineItems() != null
+          && anonymousCart.getLineItems().size() > 0) {
+        cartMerger.mergeCart(entity, anonymousCart);
+      } else {
+        LOG.debug("Anonymous cart not exist. anonymousId: {}", anonymousId);
```
Comment (davis999) [inline src/main/java/io/reactivesw/cart/application/service/CartApplication.java]: need a dot.

File: src/main/java/io/reactivesw/cart/application/service/CartApplication.java [hunk 16]
```diff
@@ -71,17 +77,58 @@ public CartView updateCart(String id, Integer version, List<UpdateAction> action
...
+      Cart anonymousCart = cartService.getCartByAnonymousId(anonymousId);
+      if (anonymousCart != null
+          && anonymousCart.getLineItems() != null
+          && anonymousCart.getLineItems().size() > 0) {
+        cartMerger.mergeCart(customerCart, anonymousCart);
+      } else {
+        LOG.debug("Anonymous cart not exist. anonymousId: {}", anonymousId);
+      }
+    }
+    LOG.debug("Exit. cart: {}", customerCart);
```
Comment (davis999) [inline src/main/java/io/reactivesw/cart/application/service/CartApplication.java]: cart should be Cart.
need a dot.

File: src/main/java/io/reactivesw/cart/application/controller/CartController.java (hunk 11 above)
Comment (kaka89) [reply src/main/java/io/reactivesw/cart/application/controller/CartController.java:66]: for all parameters we need to record,  use **CamelCase**

File: src/main/java/io/reactivesw/cart/application/controller/CartController.java (hunk 1 above)
Comment (kaka89) [reply src/main/java/io/reactivesw/cart/application/controller/CartController.java:98]: we will still get functions like: `getActiveCartByAnonymousId`

File: src/main/java/io/reactivesw/cart/application/service/CartMerger.java (hunk 8 above)
Comment (kaka89) [reply src/main/java/io/reactivesw/cart/application/service/CartMerger.java:45]: We use it in consumer.

**Pull request merged**